compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -944,6 +944,47 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         interp_ok(())
     }
 
+    pub fn prepare_for_native_call(
+        &mut self,
+        id: AllocId,
+        initial_prov: M::Provenance,
+    ) -> InterpResult<'tcx> {
+        // Expose provenance of the root allocation.
+        M::expose_provenance(self, initial_prov)?; // TODO: Is this the right way to expose provenance?
+
+        let mut done = rustc_data_structures::fx::FxHashSet::default();
+        let mut todo = vec![id];
+        while let Some(id) = todo.pop() {
+            if done.insert(id) {
```

Comment:
Please make this `if !done.insert(id) { continue; }` to avoid the rightwards drift